### PR TITLE
KAFKA-17609:[1/4] Changes needed to convert system tests to use KRaft and remove ZK 

### PR DESCRIPTION
**Pull request by bbejeck**
This is part one of a multi-pr effort to convert Kafka Streams system tests to KRaft.  I decided to break down the changes into multiple PRs to reduce the review load

- Passing system tests for [ApplicationUpgradeTest](https://semaphore.ci.confluent.io/jobs/8d9c8323-c32a-40f8-8d83-909743a86120/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976&suite_id=94b22e6d-3d1f-39cc-a647-051d783a5e26)
- Passing system tests for [BrokerBounceTest](https://semaphore.ci.confluent.io/jobs/8d9c8323-c32a-40f8-8d83-909743a86120/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976&suite_id=b320bd42-4652-31e8-a90a-cfa213509a52)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tests/kafkatest/tests/kafka_test.py
Comment (bbejeck) [inline tests/kafkatest/tests/kafka_test.py]: This is needed for now for getting the tests to pass, there will be a follow-up PR to remove

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_application_upgrade_test.py]: same here

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_application_upgrade_test.py]:
```suggestion
from kafkatest.services.kafka import KafkaService, quorum
```

File: tests/kafkatest/services/streams.py
Comment (mjsax) [inline tests/kafkatest/services/streams.py:631]: Can we remove `LATEST_0_10_0` and `LATEST_0_10_1` from the imports, too?

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: Why do we need 9 nodes now?

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: If `bounce_type` is always `full`, and `metadata_quorum` is always `quorum.combined_kraft`, why do we have both as test parameters? Can we not hard-code both inside the test itself?

Even if I don't even see where `metadata_quorum` is used? Why do we add it?

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: Would it make sense to change `KafkaService` to use `zk=None` as default so we don't have to pass it at all any longer? (Of course, touching `KafkaService` might affect _all_ system tests, so not sure what the impact / splash radius would be? -- On the other side, if `zk` is mandatory, making it optional with default `None` might not affect any existing test after all?)

Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: Why changing `num_nodes=self.replication` to `num_nodes=3` ?

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py:210]: As above: should we move this parameter and hard-code? (Similar below)

**Comment by mjsax** [comment]
Thanks Bill. Made a pass. Can you also trigger a system test run and share the link?

**Comment by bbejeck** [comment]
> Thanks Bill. Made a pass. Can you also trigger a system test run and share the link?

Passing system tests for [ApplicationUpgradeTest](https://semaphore.ci.confluent.io/jobs/8d9c8323-c32a-40f8-8d83-909743a86120/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976&suite_id=94b22e6d-3d1f-39cc-a647-051d783a5e26)
Passing system tests for [BrokerBounceTest](https://semaphore.ci.confluent.io/jobs/8d9c8323-c32a-40f8-8d83-909743a86120/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976&suite_id=b320bd42-4652-31e8-a90a-cfa213509a52)

Also in the PR description

File: tests/kafkatest/tests/kafka_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/kafka_test.py]: removed now

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: `metadata_quorum` must be in the `@matrix` annotation as it's picked up by the `ServiceQuorumInfo.for_test` class method to indicate the type of metadata quorum (via the `test_context`) - if it's not set, the test will assume you're using ZK and throws an error because the `zk` field in the `KafkaService` class is `None`.   From the git history, the team made updates some time ago adding the `metadata_quorum` parameter to streams system tests to use KRaft vs. ZK.  You'll see this pattern in all the Kafka Streams system tests.

As for `bounce_type`, I'm unfamiliar with this test; there is a condition for a "rolling" bounce type in the test, but it was never added to the test parameter list.  Do you know anything about this?  I can add "rolling" to the parameter list and exercise this portion of the test.

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: > Would it make sense to change KafkaService to use zk=None as default so we don't have to pass it at all any longer? (Of course, touching KafkaService might affect all system tests, so not sure what the impact / splash radius would be? -- On the other side, if zk is mandatory, making it optional with default None might not affect any existing test after all?)

I'll leave it as is.  the `zk` parameter has no default value in the `KafkaService` `__init__` method.  Parameters without default values must come before the ones that do have them, so to make this change will update the parameter order which I believe will affect all other system tests.

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py:210]: answered above

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: > Why changing num_nodes=self.replication to num_nodes=3 ?

It never changes, so I thought I'd set explicitly, but there's no harm in putting it back.

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: When I ran the test locally, I got a 'not enough resources' error indicating that the test needed more than 6 nodes, but  I'll set it back to 6 and run on the branch builder as it seems to have been fine before.  Probably an oversight on my part.

UPDATE: Running branch builder I received this error

```
ducktape.cluster.node_container.InsufficientResourcesError: 
Not enough nodes available to allocate. linux nodes requested: 1. linux nodes available: 0
```
So I'm reverted `num_nodes` back to 9

**Comment by bbejeck** [comment]
@mjsax all comments addressed

File: tests/setup.py
Comment (bbejeck) [inline tests/setup.py]: This change is temporary to run the system tests - I'll revert once the test completes

**Comment by bbejeck** [comment]
I kicked off a new branch builder for `streams_application_upgrade_test` and I'll post the link here when it completes.

**Comment by bbejeck** [comment]
[System test results](https://semaphore.ci.confluent.io/jobs/ce66d1c2-c5ad-453a-93d2-6577229db201/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976&suite_id=94b22e6d-3d1f-39cc-a647-051d783a5e26) 29 passed, 3 failed - I'll look into the failures.

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: I am still confuse why we need 3 more node now... no need to block the PR on this question, but I am just curious to understand why. -- I mean, if we have 3 brokers, and 3 controllers, plus 3 KS, yes, we would need 9. But before, we should have had 3 brokers, 3 ZK, and 3 KS, and thus also need 9? -- Are we actually running with dedicate quorum nodes? Or did this test not us ZK but "mixed-moded" broker/controller nodes before?

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: IIRC (but not 100% sure) `full` means we stop all nodes first, and restart them afterward, while `rolling` would be regular rolling bounce, ie, stop-start, stop-start.

Not sure if we did run this test will rolling restart in the past? If yes, maybe there was a reason to remove rolling-restart and only test with `full`? Might need some digging. -- If it details this PR, we can also do a follow up to figure this out, and either add rolling-restart or maybe remove the parameter?

Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: From the history, it looks like the `rolling` parameter was never enabled; the code is there to support the `rolling restart.  So I enabled it in this latest run, and there were three failures, all from the `rolling` portion - it's from the three oldest versions.  If debugging these failures holds up the PR too much, I'll remove `rolling` and defer it to a follow up PR.

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: Did we figure this one out?

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: So we keep as-is for now, and do a follow up PR to cleanup parameters is we don't need them or add other options?

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: Why are we removing this matrix?

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: I thought we agree to run all KS system test with `combined_kraft` ?

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py:255]: as above

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_broker_bounce_test.py:274]: as above

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: That matrix is using zookeeper

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: Maybe I got my wires crossed, I thought we wanted to go back to how they were originally, same applies to all others below.

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py]: Ah. I missed the other `@matrix` annotation above it.

Seems we can remove `metadata_quorum=[quorum.isolated_kraft]` from the remaining matrix?

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:60]: yes, I would advocate for this approach

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py:255]: Required to indicate to use KRaft for metadata quorum.  Failing to do so without a `@matrix` annotation results in this error as the test will default to ZK without it.
```File "/opt/kafka-dev/tests/kafkatest/services/kafka/kafka.py", line 541, in security_config
    zk_sasl=self.zk.zk_sasl if self.quorum_info.using_zk else False, zk_tls=self.zk_client_secure,
AttributeError: 'NoneType' object has no attribute 'zk_sasl'
```

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py:274]: Required to indicate to use KRaft for metadata quorum.  Failing to do so without a `@matrix` annotation results in this error as the test will default to ZK without it.
```File "/opt/kafka-dev/tests/kafkatest/services/kafka/kafka.py", line 541, in security_config
    zk_sasl=self.zk.zk_sasl if self.quorum_info.using_zk else False, zk_tls=self.zk_client_secure,
AttributeError: 'NoneType' object has no attribute 'zk_sasl'
```

File: tests/kafkatest/tests/streams/streams_application_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: No I have not looked into this yet, probably ok to defer this to a follow-up task

Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_application_upgrade_test.py:58]: SG. -- We should just not drop it to ensure we run system test as efficient/light weight as possible.

**Review by mjsax — APPROVED** [review]
LGTM

File: Vagrantfile
Comment (bbejeck) [inline Vagrantfile]: Temp fix until #17576 gets merged - I'll revert this if the tests pass and I can merge

File: vagrant/base.sh
Comment (bbejeck) [inline vagrant/base.sh]: Temp fix until #17576 gets merged - I'll revert this if the tests pass and I can merge

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py:274]: update: some of these tests [were failing ](https://semaphore.ci.confluent.io/jobs/f0ad9c45-d5fd-433b-a2d7-72d8e36440ff/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976) as previously configured so I needed to update to `isolated_kraft` and increase the number of nodes.



test_all_brokers_bounce_failure

File: tests/kafkatest/tests/streams/streams_broker_bounce_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_broker_bounce_test.py:255]: update: some of these tests [were failing ](https://semaphore.ci.confluent.io/jobs/f0ad9c45-d5fd-433b-a2d7-72d8e36440ff/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976) as previously configured so I needed to update to `isolated_kraft` and increase the number of nodes.


many_brokers_bounce_failures_part_1

**Comment by bbejeck** [comment]
[Link to all passing tests.](https://semaphore.ci.confluent.io/jobs/77ee30bb-55ae-4b7f-bda9-551af378e84a/summary?report_id=0356f43b-c807-318c-b61e-09e8fdd4e976)

zk_to_kraft_migration_pr_1_results

**Comment by bbejeck** [comment]
Merged #17275 into trunk

**Pull request merged**
